### PR TITLE
Adds wasi_snapshot_preview1.NewFunctionExporter

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This adds `wasi_snapshot_preview1.NewFunctionExporter` for the following two use cases:

* Overriding a builtin function with an alternate implementation.
  * Allows developers like @jerbob92 to test new WASI functions without forking wazero.
* Exporting functions to the module "wasi_unstable" for legacy code.
  * Needed by mosn to support existing wasm which may never go away.

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
cc @antJack @taoyuanyuan from mosn. This is the main change needed that lets wazero power old proxy-wasm ABI

**Comment by codefromthecrypt** [comment]
verified wazero can now implement old proxy-wasm in mosn with this https://github.com/mosn/proxy-wasm-go-host/compare/main...codefromthecrypt:wazero

**Comment by codefromthecrypt** [comment]
actually, this won't help too much with WASI because most things need internal types. anyway it would when we expose them (which isn't soon as we have to re-evaluate fs.FS)

File: imports/wasi_snapshot_preview1/usage_test.go [hunk 1]
```diff
@@ -16,18 +16,17 @@ var wasiArg []byte
...
 func TestInstantiateModule(t *testing.T) {
 	r := wazero.NewRuntime(testCtx)
+	defer r.Close(testCtx)
 
 	stdout := bytes.NewBuffer(nil)
 
 	// Configure WASI to write stdout to a buffer, so that we can verify it later.
 	sys := wazero.NewModuleConfig().WithStdout(stdout)
-	wm, err := Instantiate(testCtx, r)
+	_, err := Instantiate(testCtx, r)
```
Comment (anuraaga) [inline imports/wasi_snapshot_preview1/usage_test.go:25]: While cleaning, can consider switching to `MustInstantiate`

**Review by anuraaga — APPROVED** [review]

File: imports/wasi_snapshot_preview1/usage_test.go (hunk 1 above)
Comment (codefromthecrypt) [reply imports/wasi_snapshot_preview1/usage_test.go:25]: good idea

**Pull request merged**